content/blog/2019-09-25-function-calls.md

```diff
@@ -0,0 +1,243 @@
+
+```typescript
+<ReturnType> <function name>(<argument name> : <Type>, ...) { <instructions> };
+```
+
+Where:
+- `<ReturnType>`: The return type of a function must be `void` or one of the currently recognized Bril types: `int` or `bool`.
+- `<function name>`: The function's name.
+- `<argument name> : <Type>`: There can be zero or more arguments. Each argument name must be paired with a Bril type.
+- `<instructions>`: This is a sequence of Bril instructions.
+
+Bril now supports two kinds of `call`s, those that produce a value (value operation), and those that do not (effect operation):
+
+```typescript
+var <variable name> : <Type>  = call <function name>(<args>);
+call <function name>(<args>);
+```
+
+For backwards compatibility, functions can still be declared without return types and arguments, as in `tests/ts/br.bril`. 
+Such functions are assumed to have a return type of void.
+
+### Extended JSON representation
+
+We extended the JSON representation of Bril functions to account for a function's arguments and return type. Every `Function` object still has a name and a list of instructions. 
+
+```typescript
+{ "name": "<string>", "instrs": [<Instruction>, ...], "args": [<Argument>, ...], "type": <Type>}
+```
+
+A function can take no arguments, in which case the `\"args\"` field contains the empty list.
+The return type, represented by the `\"type\"` field, is not required. A function that does not return anything (giving it the return type `void`) does not contain the `\"type\"` field.
+
+An `Argument` JSON object contains the argument's name and type:
+
+```typescript
+{"name": "<string>", "type": <Type>}
+```
+
+The JSON Bril `Program` object remains unchanged as a list of functions.
+
+### Compiling to JSON
+
+We extended the frontend for text-based Bril in `briltxt.py`.
+The goal was to convert our new function definitions and call instructions to the JSON representation of Bril, necessitating extending the parser and JSON generators.
+
+We also extended the TypeScript frontend in `ts2bril.ts`.
+The TypeScript parser already handled calls to handle treating `console.log` statements as Bril's `print` statements. We extended this component to also capture effectful calls that return results. In addition, the initial implementation did not support function clarations, so we added new support to transform the declaration and type information.
+
+### Interpreter
+
+The interpreter needed to be able to handle functions and calls in their extended JSON representation. 
+The main work was when encountering a `call` instruction: we create a new, empty environment with the arguments bound to the correct values.
+The interpreter searches for the function name in the program's list of functions since we are not implementing first-class functions.
+Because we chose to represent the stack implicitly, function calls are executed with a recursive call to `evalInstr`, thus relying on the underlying TypeScript stack frame implementation.
```

Comment:
Just for fun, you might be interested in @rolph-recto's implementation (#32), which has an explicit stack to help implement exceptions.